Adds region selection to TsunamiBuilder

simonrw opened the pull request:
Implement custom region selection for Tsunami

This goes half way to solving #3. It allows the user to select a
different aws region from us-east-1.

The documentation for the method includes a link to the
`rusoto_core::Region` struct, and displays the default value.

simonrw committed:
Adds region selection to TsunamiBuilder

simonrw commented:
There is the question of: is this the desired behaviour? Should the code default to `us-east-1`, or error when the region is not supplied? It seems quite a fundamental piece of information from the user, that perhaps it makes sense making it a critical error if it's _not_ supplied?

jonhoo commented:
This looks good! I think it'd also be nice to include a link to https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-available-regions.

As far as your question goes: yes, it's not entirely clear that any one region is a more sensible default than the others. I think the right thing here (and probably in general) is to provide convenience rather than forcing the user to make a choice, even when there is no good default value. One way to make this choice is to think about the downsides of the default being wrong. If the downsides are severe, then it might make sense to not have a default and force a choice (i.e., by erroring if it's not given). In our case though, the downsides are fairly minor (ssh latency will be a bit higher), except in the case where the instances need to talk to some *particular* third party, where the user explicitly selecting a region seems like a reasonable cost.

simonrw committed:
Include link to amazons region descriptions

simonrw committed:
Rename region method

The builder pattern currently uses the "set_" convention, thus using the
methods on an already constructed object:

```
let b = Builder::new();
b.set_this();
b.set_that();
b.run();
```

rather than

```
let b = Builder::new()
    .with_this()
    .with_that()
    .run();
```

This was discussed during the screencast, so we try to stick to this
convention.

simonrw commented:
I've updated the docs, and renamed the method to `set_region` which fits in more with the naming convention of the other methods.

I think you're right with the lack of defaults. My only concern is unexpected cost differences, but I don't think AWS pricing varies that much between regions.

jonhoo commented:
If we move to a builder pattern more in line with [`C-BUILDER` in the Rust API guidelines](https://rust-lang-nursery.github.io/api-guidelines/type-safety.html#c-builder) then we'll probably pick more ergonomic method names, but I think `set_*` is fine for now. Thanks!

jonhoo commented:
@mindriot101 see also https://github.com/rusoto/rusoto/issues/957
